src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs

```diff
@@ -52,60 +55,80 @@ internal abstract partial class AbstractNavigateToSearchService
             var lowPriDocs = project.Documents.Where(d => !highPriDocs.Contains(d)).ToSet();
-            await ProcessDocumentsAsync(searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, lowPriDocs, cancellationToken).ConfigureAwait(false);
+            await foreach (var item in ProcessDocumentsAsync(searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, lowPriDocs, cancellationToken).WithCancellation(cancellationToken))
+                yield return item;
         }
 
-        private static async Task ProcessDocumentsAsync(
+        private static IAsyncEnumerable<RoslynNavigateToItem> ProcessDocumentsAsync(
             Document? searchDocument,
             string patternName,
             string? patternContainer,
             DeclaredSymbolInfoKindSet kinds,
-            Func<RoslynNavigateToItem, Task> onResultFound,
             ISet<Document> documents,
             CancellationToken cancellationToken)
         {
-            using var _ = ArrayBuilder<Task>.GetInstance(out var tasks);
+            using var _ = ArrayBuilder<IAsyncEnumerable<RoslynNavigateToItem>>.GetInstance(out var builders);
 
             foreach (var document in documents)
             {
                 if (searchDocument != null && searchDocument != document)
                     continue;
 
                 cancellationToken.ThrowIfCancellationRequested();
-                tasks.Add(Task.Run(() =>
-                    ProcessDocumentAsync(document, patternName, patternContainer, kinds, onResultFound, cancellationToken), cancellationToken));
+                builders.Add(ProcessDocumentAsync(document, patternName, patternContainer, kinds, cancellationToken));
             }
 
-            await Task.WhenAll(tasks).ConfigureAwait(false);
+            return builders.ToArray().MergeAsync(cancellationToken);
         }
 
-        private static async Task ProcessDocumentAsync(
+        private static async IAsyncEnumerable<RoslynNavigateToItem> ProcessDocumentAsync(
             Document document,
             string patternName,
             string? patternContainer,
             DeclaredSymbolInfoKindSet kinds,
-            Func<RoslynNavigateToItem, Task> onResultFound,
-            CancellationToken cancellationToken)
+            [EnumeratorCancellation] CancellationToken cancellationToken)
         {
             var index = await TopLevelSyntaxTreeIndex.GetRequiredIndexAsync(document, cancellationToken).ConfigureAwait(false);
 
-            await ProcessIndexAsync(
-                document.Id, document, patternName, patternContainer, kinds, onResultFound, index, cancellationToken).ConfigureAwait(false);
+            await foreach (var item in ProcessIndexAsync(
+                document.Id, document, patternName, patternContainer, kinds, index, cancellationToken).WithCancellation(cancellationToken))
+            {
+                yield return item;
+            }
```

Comment:
would love a way to directly return an IAsyncEnumerable from within an `*async*` IAsyncEnumerable.  @jcouv anything better that can be done here?

Reply:
You could write a regular async method (as opposed to an async iterator method):
```
async Task<IAsyncEnumerable<RoslynNavigateToItem>> ProcessDocumentAsync(...)
{
  var index = await TopLevelSyntaxTreeIndex.GetRequiredIndexAsync(document, cancellationToken).ConfigureAwait(false);
  return ProcessIndexAsync(...);
}
```